### PR TITLE
[k8s] Add `sky status` flag to query global Kubernetes status

### PR DESCRIPTION
**Pull request by romilbhardwaj**
Adds a `--kubernetes` flag to `sky status` to show the global state of the Kubernetes cluster, including SkyPilot clusters created by other users. Helps users see the current state of the Kubernetes cluster.

Example:

```
(base) ➜  sky-experiments git:(k8s_global_status) ✗ sky status --kubernetes
SkyPilot Clusters on Kubernetes
NAME        USER    LAUNCHED     RESOURCES                                      STATUS  
test-2ea4   romilb  45 secs ago  2x Kubernetes(cpus=2.0, mem=2.0)               UP      
train-j49a  test    1 min ago    2x Kubernetes(cpus=2.0, mem=8.0, {'H100': 1})  UP     
```

TODO:
- [x] Handle multiple contexts
- [x] Decide whether to poll current namespace or all namespaces
- [x] Remove user hash from cluster name
- [x] Cleaner code

Tested (run the relevant ones):

- [ ] Code formatting: `bash format.sh`
- [ ] Any manual or new tests for this PR (please specify below)
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `conda deactivate; bash -i tests/backward_compatibility_tests.sh`

### CODE REVIEW

**Comment by romilbhardwaj** [comment]
Updated to query job status from all running controllers:

```
(base) ➜  sky-experiments git:(k8s_global_status) ✗ sky status --kubernetes
SkyPilot Clusters on Kubernetes
USER     NAME                               LAUNCHED        RESOURCES                          STATUS  
gcpuser  sky-cmd-2-80b5                     9 mins ago      1x Kubernetes(cpus=2.0, mem=2.0)   UP      
romilb   sky-cmd-3-2ea4                     19 mins ago     1x Kubernetes(cpus=2.0, mem=2.0)   UP      
romilb   sky-jobs-controller-2ea485ea-2ea4  4 hrs ago       1x Kubernetes(cpus=4.0, mem=4.0)   UP      
gcpuser  sky-jobs-controller-80b50983-80b5  3 hrs ago       1x Kubernetes(cpus=8.0, mem=24.0)  UP      
romilb   train-2ea4                         a few secs ago  1x Kubernetes(cpus=2.0, mem=2.0)   UP      
Managed jobs from 2 users on Kubernetes
In progress tasks: 2 RUNNING, 1 STARTING
USER     ID  TASK  NAME     RESOURCES   SUBMITTED    TOT. DURATION  JOB DURATION  #RECOVERIES  STATUS     
romilb   4   -     sky-cmd  1x[CPU:16]  8 mins ago   8m 14s         -             0            STARTING   
romilb   3   -     sky-cmd  1x[CPU:1+]  19 mins ago  19m 18s        18m 31s       0            RUNNING    
romilb   2   -     sky-cmd  1x[CPU:1+]  1 hr ago     1h 1m 2s       1h 16s        0            SUCCEEDED  
romilb   1   -     sky-cmd  1x[CPU:1+]  4 hrs ago    1m 32s         8s            0            SUCCEEDED  
gcpuser  2   -     sky-cmd  1x[CPU:1+]  9 mins ago   9m 57s         9m 14s        0            RUNNING    
gcpuser  1   -     sky-cmd  1x[CPU:1+]  3 hrs ago    49s            8s            0            SUCCEEDED  
```

Still needs clean up and edge case handling.

**Comment by romilbhardwaj** [comment]
This should be ready for review now. 
skystatus
- [ ] Handle sky serve clusters?

**Comment by romilbhardwaj** [comment]
Few nits, added hint for SkyServe replicas and `--k8s` shorthand:

image

**Comment by concretevitamin** [comment]
UX LGTM; quick nits:
- I'd expect the spinner doesn't refresh for every user name, esp. when there are a lot of users?
- `show-gpus` context is None while `status --k8s` shows non-None context
  - also: former has `Context`, latter is lower case
- `-h: [Experimental] Show SkyPilot clusters from all users on Kubernetes.` --> `[Experimental] Show all SkyPilot resources (including from other users) of the current Kubernetes context.`?
- ` --k8s`: discussed offline, this seems like a name we're ok with for now
- `Hint: SkyServe controllers detected in the cluster. SkyServe service replicas will be shown as SkyPilot clusters.`
  - `Hint: Currently, SkyServe replica pods are shown in the "SkyPilot clusters" section.`
  - Maybe dim it since it doesn't warrant a lot of attention.
- `Managed jobs from 2 users` -> `Managed jobs`

I tried launching a managed job on the same shared k8s cluster, and the job loops forever in starting. Controller logs:
```
...
(sky-cmd, pid=4467) Creating a new cluster: 'sky-cmd-1' [1x Kubernetes(2CPU--2GB, cpus=2)].
(sky-cmd, pid=4467) Tip: to reuse an existing cluster, specify --cluster (-c). Run `sky status` to see existing clusters.
(sky-cmd, pid=4467) To view detailed progress: tail -n100 -f /home/gcpuser/sky_logs/sky-2024-10-09-23-21-21-787148/provision.log
(sky-cmd, pid=4467) Launching on Kubernetes 'sky-cmd-1'.
(sky-cmd, pid=4467) run_instances: Error occurred when creating pods: Failed to check user privilege for pod sky-cmd-1-8a39-head with return code 1: 'if [ $(id -u) -eq 0 ]; then  echo \'alias sudo=""\' >> ~/.bashrc; echo succeed;else   if command -v sudo >/dev/null 2>&1; then     timeout 2 sudo -l >/dev/null 2>&1 && echo succeed ||     ( echo 52; );   else     ( echo 52; );   fi; fi'
(sky-cmd, pid=4467) Output: error: The gcp auth plugin has been removed.
(sky-cmd, pid=4467) Please use the "gke-gcloud-auth-plugin" kubectl/client-go credential plugin instead.
(sky-cmd, pid=4467) See https://cloud.google.com/blog/products/containers-kubernetes/kubectl-auth-changes-in-gke for further details
(sky-cmd, pid=4467) .
(sky-cmd, pid=4467) sky.exceptions.ResourcesUnavailableError: Failed to acquire resources in all zones in gke_skypilot-375900_us-central1-c_gkeusc6. Try changing resource requirements or use another region.
...
```

**Comment by romilbhardwaj** [comment]
Fixed UX comments and added error handling.

File: sky/cli.py [hunk 1]
```diff
@@ -1458,6 +1458,73 @@ def _get_services(service_names: Optional[List[str]],
     return num_services, msg
 
 
+def _status_kubernetes(show_all: bool):
```
Comment (concretevitamin) [inline sky/cli.py:1461]: Quick 1-line docstr?

File: sky/utils/cli_utils/status_utils.py [hunk 2]
```diff
@@ -316,3 +301,143 @@ def _get_estimated_cost_for_cost_report(
...
+                row.append(status_column.calc(cluster))
+        cluster_table.add_row(row)
+
+    if clusters:
+        click.echo(f'{colorama.Fore.CYAN}{colorama.Style.BRIGHT}'
+                   f'SkyPilot clusters'
+                   f'{colorama.Style.RESET_ALL}')
+        click.echo(cluster_table)
+    else:
+        click.echo('No existing SkyPilot clusters on Kubernetes.')
```
Comment (concretevitamin) [inline sky/utils/cli_utils/status_utils.py:345]: "No SkyPilot resources found in the active Kubernetes context."

(not just clusters)

File: sky/data/storage_utils.py [hunk 3]
```diff
@@ -22,6 +21,8 @@
     'to the cloud storage for {path!r}'
     'due to the following error: {error_msg!r}')
 
+LAST_USE_TRUNC_LENGTH = 25
```
Comment (concretevitamin) [inline sky/data/storage_utils.py:24]: - prepend `_`?

File: sky/jobs/core.py [hunk 4]
```diff
@@ -138,6 +140,81 @@ def launch(
                    _disable_controller_check=True)
 
 
+def queue_kubernetes(pod_name: str,
+                     context: Optional[str] = None,
+                     skip_finished: bool = False) -> List[Dict[str, Any]]:
+    """Gets the jobs queue from a specific controller pod.
```
Comment (concretevitamin) [inline sky/jobs/core.py:146]: This naming is surprising. From the name I thought it's "gets the queue info for an entire k8s cluster". Maybe rename?

File: sky/utils/cli_utils/status_utils.py [hunk 5]
```diff
@@ -316,3 +301,143 @@ def _get_estimated_cost_for_cost_report(
...
+
+
+def process_skypilot_pods(
+    pods: List[Any],
+    context: Optional[str] = None
+) -> Tuple[List[Dict[Any, Any]], Dict[str, Any], Dict[str, Any]]:
```
Comment (concretevitamin) [inline sky/utils/cli_utils/status_utils.py:351]: quick docstr?

File: sky/utils/common_utils.py [hunk 6]
```diff
@@ -679,3 +679,22 @@ def new_func(*args, **kwargs):
         return func(*args, **kwargs)
 
     return new_func
+
+
+def truncate_long_string(s: str, max_length: int = 35) -> str:
```
Comment (concretevitamin) [inline sky/utils/common_utils.py:684]: quick docstr

**Review by concretevitamin — APPROVED** [review]
LGTM, thanks @romilbhardwaj.

File: sky/jobs/core.py (hunk 4 above)
Comment (romilbhardwaj) [reply sky/jobs/core.py]: Updated to `queue_from_kubernetes_pod`

**Pull request merged**